src/front/glsl/ast.rs

```diff
@@ -141,99 +142,112 @@ pub enum Profile {
-        Context {
+        let mut this = Context {
             expressions,
             locals,
             arguments,
 
             scopes: vec![FastHashMap::default()],
-            lookup_global_var_exps: FastHashMap::default(),
-            lookup_constant_exps: FastHashMap::default(),
+            lookup_global_var_exps: FastHashMap::with_capacity_and_hasher(
+                program.constants.len() + program.global_variables.len(),
+                Default::default(),
+            ),
             typifier: Typifier::new(),
             samplers: FastHashMap::default(),
 
             hir_exprs: Arena::default(),
-        }
-    }
+            emitter: Emitter::default(),
+        };
 
-    pub fn lookup_local_var(&self, name: &str) -> Option<VariableReference> {
-        for scope in self.scopes.iter().rev() {
-            if let Some(var) = scope.get(name) {
-                return Some(var.clone());
-            }
-        }
-        None
-    }
+        this.emit_start();
 
-    pub fn lookup_global_var(
-        &mut self,
-        program: &mut Program,
-        name: &str,
-    ) -> Option<VariableReference> {
-        self.lookup_global_var_exps.get(name).cloned().or_else(|| {
-            let expr = match *program.lookup_global_variables.get(name)? {
+        for &(ref name, lookup) in program.global_variables.iter() {
+            let expr = match lookup {
                 GlobalLookup::Variable(v) => Expression::GlobalVariable(v),
                 GlobalLookup::BlockSelect(handle, index) => {
-                    let base = self.expressions.append(Expression::GlobalVariable(handle));
+                    let base = this.add_expression(Expression::GlobalVariable(handle), body);
 
                     Expression::AccessIndex { base, index }
                 }
             };
 
-            let expr = self.expressions.append(expr);
+            let expr = this.add_expression(expr, body);
             let var = VariableReference {
                 expr,
-                load: Some(self.expressions.append(Expression::Load { pointer: expr })),
+                load: Some(this.add_expression(Expression::Load { pointer: expr }, body)),
                 // TODO: respect constant qualifier
                 mutable: true,
             };
 
-            self.lookup_global_var_exps.insert(name.into(), var.clone());
-
-            Some(var)
-        })
-    }
-
-    pub fn lookup_constants_var(
-        &mut self,
-        program: &mut Program,
-        name: &str,
-    ) -> Option<VariableReference> {
-        self.lookup_constant_exps.get(name).cloned().or_else(|| {
-            let expr = self
-                .expressions
-                .append(Expression::Constant(*program.lookup_constants.get(name)?));
+            this.lookup_global_var_exps.insert(name.into(), var);
+        }
 
+        for &(ref name, handle) in program.constants.iter() {
+            let expr = this.add_expression(Expression::Constant(handle), body);
```

Comment:
similarly, here we know that none of these expressions need to be explicitly emitted

Reply:
Yes but the previous iteration might have `Load` to be emitted but a flush before beginning the loop can fix this

Reply:
sure, you can also just move this loop above the globals loop